File: ex13/DeadPixelDetector.cpp
#include <opencv2/opencv.hpp>
#include <nlohmann/json.hpp>
#include <vector>
#include <unordered_map>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <cmath>
#include <memory>

using json = nlohmann::json;

// Structure to store pixel defect information
struct PixelDefect {
    int x;
    int y;
    double confidence;
    int start_frame;
    int end_frame;
    int total_appearances;

    PixelDefect(int x_, int y_, double conf_, int start_)
        : x(x_), y(y_), confidence(conf_), start_frame(start_),
          end_frame(-1), total_appearances(1) {}

    json to_json() const {
        return {
            {"x", x},
            {"y", y},
            {"confidence", std::round(confidence * 1000.0) / 1000.0},
            {"start_frame", start_frame},
            {"end_frame", end_frame},
            {"total_appearances", total_appearances},
            {"span", end_frame >= 0 ? end_frame - start_frame : nullptr}
        };
    }
};

// Custom hash function for pixel coordinates
struct CoordHash {
    size_t operator()(const std::pair<int, int>& p) const {
        return std::hash<int>()(p.first) ^ (std::hash<int>()(p.second) << 1);
    }
};

class DeadPixelDetector {
public:
    DeadPixelDetector(double distance_threshold = 3.0,
                      int min_persistence = 5,
                      double noise_threshold = 0.1)
        : distance_threshold_(distance_threshold),
          min_persistence_(min_persistence),
          noise_threshold_(noise_threshold) {}

    void process_video(const std::string& input_path,
                      const std::string& json_output_path,
                      const std::string& video_output_path = "") {
        cv::VideoCapture cap(input_path);
        if (!cap.isOpened()) {
            throw std::runtime_error("Could not open input video: " + input_path);
        }

        std::unique_ptr<cv::VideoWriter> out;
        if (!video_output_path.empty()) {
            int fourcc = cv::VideoWriter::fourcc('X', 'V', 'I', 'D');
            double fps = cap.get(cv::CAP_PROP_FPS);
            cv::Size frame_size(
                static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH)),
                static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT))
            );
            out = std::make_unique<cv::VideoWriter>(
                video_output_path, fourcc, fps, frame_size
            );
        }

        int frame_count = 0;
        int total_frames = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_COUNT));
        cv::Mat frame, gray_frame;

        std::cout << "Processing video: " << input_path << std::endl;

        while (true) {
            cap >> frame;
            if (frame.empty()) break;

            cv::cvtColor(frame, gray_frame, cv::COLOR_BGR2GRAY);
            auto current_defects = detect_defective_pixels(gray_frame);
            update_defects(current_defects, frame_count);

            if (out) {
                visualize_defects(frame);
                *out << frame;
            }

            frame_count++;
            if (frame_count % 100 == 0) {
                std::cout << "Processed frame " << frame_count << "/"
                         << total_frames << std::endl;
            }
        }

        save_results(json_output_path);
    }

private:
    double distance_threshold_;
    int min_persistence_;
    double noise_threshold_;
    std::unordered_map<std::pair<int, int>, PixelDefect, CoordHash> active_defects_;
    std::vector<PixelDefect> completed_defects_;

    std::vector<cv::Mat> compute_difference_matrices(const cv::Mat& frame) {
        static const std::vector<std::pair<int, int>> directions = {
            {1, 1}, {1, 0}, {1, -1}, {0, -1},
            {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}
        };

        std::vector<cv::Mat> diff_matrices;
        diff_matrices.reserve(directions.size());

        for (const auto& [dx, dy] : directions) {
            cv::Mat shifted;
            cv::Mat translation = (cv::Mat_<double>(2, 3) <<
                1, 0, dx,
                0, 1, dy);
            cv::warpAffine(frame, shifted, translation, frame.size(),
                          cv::INTER_LINEAR, cv::BORDER_REPLICATE);
            
            cv::Mat diff;
            cv::absdiff(frame, shifted, diff);
            diff_matrices.push_back(diff);
        }

        return diff_matrices;
    }

    std::pair<cv::Mat, cv::Mat> compute_gradient_matrices(
        const std::vector<cv::Mat>& diff_matrices) {
        
        std::vector<cv::Mat> first_channels(diff_matrices.begin(),
                                          diff_matrices.begin() + 4);
        cv::Mat first_degree;
        cv::merge(first_channels, first_degree);

        cv::Mat second_degree;
        cv::merge(diff_matrices, second_degree);

        // Apply Gaussian smoothing
        for (int i = 0; i < first_degree.channels(); ++i) {
            cv::Mat channel;
            cv::extractChannel(first_degree, channel, i);
            cv::GaussianBlur(channel, channel, cv::Size(3, 3), 0.5);
            cv::insertChannel(channel, first_degree, i);
        }

        for (int i = 0; i < second_degree.channels(); ++i) {
            cv::Mat channel;
            cv::extractChannel(second_degree, channel, i);
            cv::GaussianBlur(channel, channel, cv::Size(3, 3), 0.5);
            cv::insertChannel(channel, second_degree, i);
        }

        return {first_degree, second_degree};
    }

    double compute_mahalanobis_distance(const cv::Mat& vector,
                                      const cv::Mat& mean,
                                      const cv::Mat& icov) {
        cv::Mat diff = vector - mean;
        cv::Mat mult = diff * icov * diff.t();
        return std::sqrt(mult.at<double>(0, 0));
    }

    std::vector<PixelDefect> detect_defective_pixels(const cv::Mat& frame) {
        auto diff_matrices = compute_difference_matrices(frame);
        auto [first_grad, second_grad] = compute_gradient_matrices(diff_matrices);

        std::vector<PixelDefect> candidates;
        const int border = 2;

        for (int y = border; y < frame.rows - border; ++y) {
            for (int x = border; x < frame.cols - border; ++x) {
                cv::Mat first_grad_vector(1, first_grad.channels(), CV_64F);
                cv::Mat second_grad_vector(1, second_grad.channels(), CV_64F);

                // Extract gradient vectors
                for (int c = 0; c < first_grad.channels(); ++c) {
                    first_grad_vector.at<double>(0, c) =
                        first_grad.at<cv::Vec<uchar, 4>>(y, x)[c];
                }
                for (int c = 0; c < second_grad.channels(); ++c) {
                    second_grad_vector.at<double>(0, c) =
                        second_grad.at<cv::Vec<uchar, 8>>(y, x)[c];
                }

                if (cv::mean(first_grad_vector)[0] < noise_threshold_)
                    continue;

                try {
                    cv::Mat covar, mean;
                    cv::calcCovarMatrix(second_grad_vector, covar, mean,
                                      cv::COVAR_NORMAL | cv::COVAR_ROWS);
                    
                    if (cv::determinant(covar) < 1e-10)
                        continue;

                    cv::Mat icov = covar.inv();
                    double distance = compute_mahalanobis_distance(
                        first_grad_vector, mean, icov);

                    if (distance > distance_threshold_) {
                        candidates.emplace_back(x, y, distance, 0);
                    }
                }
                catch (const cv::Exception&) {
                    continue;
                }
            }
        }

        return candidates;
    }

    void update_defects(const std::vector<PixelDefect>& current_defects,
                       int frame_number) {
        std::unordered_set<std::pair<int, int>, CoordHash> current_coords;

        // Update active defects
        for (const auto& current : current_defects) {
            std::pair<int, int> coord{current.x, current.y};
            current_coords.insert(coord);

            auto it = active_defects_.find(coord);
            if (it != active_defects_.end()) {
                it->second.total_appearances++;
                it->second.confidence = std::max(
                    it->second.confidence, current.confidence);
            }
            else {
                PixelDefect defect = current;
                defect.start_frame = frame_number;
                active_defects_[coord] = defect;
            }
        }

        // Check for ended defects
        std::vector<std::pair<int, int>> to_remove;
        for (const auto& [coord, defect] : active_defects_) {
            if (current_coords.find(coord) == current_coords.end()) {
                if (defect.total_appearances >= min_persistence_) {
                    PixelDefect completed = defect;
                    completed.end_frame = frame_number - 1;
                    completed_defects_.push_back(completed);
                }
                to_remove.push_back(coord);
            }
        }

        for (const auto& coord : to_remove) {
            active_defects_.erase(coord);
        }
    }

    void visualize_defects(cv::Mat& frame) {
        for (const auto& [coord, defect] : active_defects_) {
            double confidence_color = std::min(defect.confidence / 10.0, 1.0);
            cv::Scalar color(
                0,
                static_cast<int>(255 * (1 - confidence_color)),
                static_cast<int>(255 * confidence_color)
            );
            cv::circle(frame, cv::Point(defect.x, defect.y), 2, color, -1);
        }
    }

    void save_results(const std::string& output_path) {
        // Finalize remaining active defects
        for (const auto& [coord, defect] : active_defects_) {
            if (defect.total_appearances >= min_persistence_) {
                PixelDefect completed = defect;
                completed.end_frame = completed.start_frame +
                                    completed.total_appearances - 1;
                completed_defects_.push_back(completed);
            }
        }

        // Create JSON output
        json result;
        result["metadata"]["timestamp"] = get_current_timestamp();
        result["metadata"]["settings"] = {
            {"distance_threshold", distance_threshold_},
            {"min_persistence", min_persistence_},
            {"noise_threshold", noise_threshold_}
        };

        result["defects"] = json::array();
        for (const auto& defect : completed_defects_) {
            result["defects"].push_back(defect.to_json());
        }

        // Save to file
        std::ofstream out(output_path);
        out << std::setw(2) << result << std::endl;

        std::cout << "Results saved to " << output_path << std::endl;
        std::cout << "Total dead pixels detected: "
                 << completed_defects_.size() << std::endl;
    }

    std::string get_current_timestamp() {
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        std::stringstream ss;
        ss << std::put_time(std::localtime(&time), "%Y-%m-%dT%H:%M:%S");
        return ss.str();
    }
};

int main(int argc, char** argv) {
    try {
        DeadPixelDetector detector(3.0, 5, 0.1);
        detector.process_video(
            "input_video.mp4",
            "dead_pixels_report.json",
            "output_video.mp4"  // Optional
        );
    }
    catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: ex13/DeadPixelDetector.py
import cv2
import numpy as np
from scipy.spatial.distance import mahalanobis
from typing import List, Tuple, Optional, Dict
from dataclasses import dataclass, asdict
import logging
from pathlib import Path
import json
from datetime import datetime

@dataclass
class PixelDefect:
    """Data class to store information about detected pixel defects."""
    x: int
    y: int
    confidence: float
    start_frame: int
    end_frame: Optional[int] = None
    total_appearances: int = 1

    def to_dict(self) -> Dict:
        """Convert defect information to dictionary format."""
        return {
            'x': self.x,
            'y': self.y,
            'confidence': round(float(self.confidence), 3),
            'start_frame': self.start_frame,
            'end_frame': self.end_frame,
            'total_appearances': self.total_appearances,
            'span': self.end_frame - self.start_frame if self.end_frame else None
        }

class DeadPixelDetector:
    def __init__(self, 
                 distance_threshold: float = 3.0,
                 min_persistence: int = 5,
                 noise_threshold: float = 0.1):
        """
        Initialize the dead pixel detector.
        
        Args:
            distance_threshold: Mahalanobis distance threshold for defect detection
            min_persistence: Minimum number of frames a defect must persist
            noise_threshold: Threshold for filtering out noise
        """
        self.distance_threshold = distance_threshold
        self.min_persistence = min_persistence
        self.noise_threshold = noise_threshold
        self.active_defects: Dict[Tuple[int, int], PixelDefect] = {}
        self.completed_defects: List[PixelDefect] = []
        self._setup_logging()

    def _setup_logging(self):
        """Set up logging configuration."""
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s'
        )
        self.logger = logging.getLogger(__name__)

    def compute_difference_matrices(self, frame: np.ndarray) -> List[np.ndarray]:
        """Generate 8 difference matrices for the given frame."""
        directions = [
            (1, 1), (1, 0), (1, -1), (0, -1),
            (-1, -1), (-1, 0), (-1, 1), (0, 1)
        ]
        diff_matrices = []
        
        for dx, dy in directions:
            shifted_frame = np.roll(frame, shift=(dx, dy), axis=(0, 1))
            if dx > 0:
                shifted_frame[0, :] = frame[0, :]
            elif dx < 0:
                shifted_frame[-1, :] = frame[-1, :]
            if dy > 0:
                shifted_frame[:, 0] = frame[:, 0]
            elif dy < 0:
                shifted_frame[:, -1] = frame[:, -1]
                
            diff = cv2.absdiff(frame, shifted_frame)
            diff_matrices.append(diff)
            
        return diff_matrices

    def compute_gradient_matrices(self, diff_matrices: List[np.ndarray]) -> Tuple[np.ndarray, np.ndarray]:
        """Compute first and second degree gradient matrices."""
        first_degree = np.stack(diff_matrices[:4], axis=-1)
        second_degree = np.stack(diff_matrices, axis=-1)
        
        for i in range(first_degree.shape[-1]):
            first_degree[..., i] = cv2.GaussianBlur(first_degree[..., i], (3, 3), 0.5)
        for i in range(second_degree.shape[-1]):
            second_degree[..., i] = cv2.GaussianBlur(second_degree[..., i], (3, 3), 0.5)
            
        return first_degree, second_degree

    def detect_defective_pixels(self, frame: np.ndarray) -> List[PixelDefect]:
        """Detect defective pixels in the frame."""
        diff_matrices = self.compute_difference_matrices(frame)
        first_grad, second_grad = self.compute_gradient_matrices(diff_matrices)
        
        candidates = []
        h, w = frame.shape
        
        y_coords, x_coords = np.mgrid[2:h-2, 2:w-2]
        
        for y, x in zip(y_coords.flatten(), x_coords.flatten()):
            first_grad_vector = first_grad[y, x, :].flatten()
            second_grad_vector = second_grad[y, x, :].flatten()
            
            if np.mean(first_grad_vector) < self.noise_threshold:
                continue
                
            try:
                covariance_matrix = np.cov(second_grad_vector, rowvar=False)
                if np.linalg.det(covariance_matrix) < 1e-10:
                    continue
                
                mean_diff = first_grad_vector - np.mean(second_grad_vector)
                distance = mahalanobis(
                    mean_diff,
                    np.mean(second_grad_vector),
                    np.linalg.inv(covariance_matrix)
                )
                
                if distance > self.distance_threshold:
                    candidates.append(PixelDefect(x=x, y=y, confidence=distance, start_frame=0))
                    
            except np.linalg.LinAlgError:
                continue
                
        return candidates

    def update_defects(self, current_defects: List[PixelDefect], frame_number: int):
        """Update tracking of dead pixels across frames."""
        # Check current defects against active defects
        current_coords = set()
        for current in current_defects:
            matched = False
            current_coord = (current.x, current.y)
            current_coords.add(current_coord)
            
            if current_coord in self.active_defects:
                defect = self.active_defects[current_coord]
                defect.total_appearances += 1
                defect.confidence = max(defect.confidence, current.confidence)
                matched = True
                
            if not matched:
                current.start_frame = frame_number
                self.active_defects[current_coord] = current

        # Check for ended defects
        ended_coords = []
        for coord, defect in self.active_defects.items():
            if coord not in current_coords:
                if defect.total_appearances >= self.min_persistence:
                    defect.end_frame = frame_number - 1
                    self.completed_defects.append(defect)
                ended_coords.append(coord)
                
        for coord in ended_coords:
            del self.active_defects[coord]

    def save_results(self, output_path: str):
        """Save detection results to JSON file."""
        # Finalize any remaining active defects
        for defect in self.active_defects.values():
            if defect.total_appearances >= self.min_persistence:
                defect.end_frame = defect.start_frame + defect.total_appearances - 1
                self.completed_defects.append(defect)

        # Prepare results
        results = {
            'metadata': {
                'timestamp': datetime.now().isoformat(),
                'settings': {
                    'distance_threshold': self.distance_threshold,
                    'min_persistence': self.min_persistence,
                    'noise_threshold': self.noise_threshold
                }
            },
            'defects': [defect.to_dict() for defect in self.completed_defects]
        }

        # Save to file
        with open(output_path, 'w') as f:
            json.dump(results, f, indent=2)
            
        self.logger.info(f"Results saved to {output_path}")
        self.logger.info(f"Total dead pixels detected: {len(self.completed_defects)}")

    def process_video(self,
                     input_path: str,
                     json_output_path: str,
                     video_output_path: Optional[str] = None):
        """
        Process video and detect dead pixels.
        
        Args:
            input_path: Path to input video
            json_output_path: Path to save JSON results
            video_output_path: Optional path to save visualization video
        """
        input_path = Path(input_path)
        if not input_path.exists():
            raise FileNotFoundError(f"Input video not found: {input_path}")
            
        cap = cv2.VideoCapture(str(input_path))
        out = None
        
        if video_output_path:
            fourcc = cv2.VideoWriter_fourcc(*'XVID')
            fps = int(cap.get(cv2.CAP_PROP_FPS))
            width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
            height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
            out = cv2.VideoWriter(video_output_path, fourcc, fps, (width, height))
            
        frame_count = 0
        total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
        
        self.logger.info(f"Processing video: {input_path}")
        
        while cap.isOpened():
            ret, frame = cap.read()
            if not ret:
                break
                
            gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
            current_defects = self.detect_defective_pixels(gray_frame)
            self.update_defects(current_defects, frame_count)
            
            if video_output_path:
                # Visualize active and completed defects
                for defect in self.active_defects.values():
                    confidence_color = np.clip(defect.confidence / 10.0, 0, 1)
                    color = (
                        0,
                        int(255 * (1 - confidence_color)),
                        int(255 * confidence_color)
                    )
                    cv2.circle(frame, (defect.x, defect.y), 2, color, -1)
                out.write(frame)
            
            frame_count += 1
            if frame_count % 100 == 0:
                self.logger.info(f"Processed frame {frame_count}/{total_frames}")
                
        cap.release()
        if out:
            out.release()
        cv2.destroyAllWindows()
        
        # Save results to JSON
        self.save_results(json_output_path)

def main():
    """Example usage of the DeadPixelDetector."""
    detector = DeadPixelDetector(
        distance_threshold=3.0,
        min_persistence=5,
        noise_threshold=0.1
    )
    
    try:
        detector.process_video(
            input_path="input_video.mp4",
            json_output_path="dead_pixels_report.json",
            video_output_path="output_video.mp4"  # Optional
        )
    except Exception as e:
        logging.error(f"Error processing video: {e}")

if __name__ == "__main__":
    main()